src/solver/mod.rs

```diff
@@ -1195,46 +1246,56 @@ impl<D: DependencyProvider, RT: AsyncRuntime> Solver<D, RT> {
             let decided = self
                 .decision_tracker
                 .try_add_decision(
-                    Decision::new(literal.solvable_id, decision, clause_id),
+                    Decision::new(literal.solvable_id(), decision, clause_id),
                     level,
                 )
                 .map_err(|_| {
-                    PropagationError::Conflict(literal.solvable_id, decision, clause_id)
+                    PropagationError::Conflict(literal.solvable_id(), decision, clause_id)
                 })?;
 
             if decided {
                 tracing::trace!(
                     "├─ Propagate assertion {} = {}",
-                    literal.solvable_id.display(self.provider()),
+                    literal.solvable_id().display(self.provider()),
                     decision
                 );
             }
         }
 
         // Watched solvables
+        let clauses = self.clauses.kinds.borrow();
+        let mut clause_states = self.clauses.states.borrow_mut();
         while let Some(decision) = self.decision_tracker.next_unpropagated() {
-            let pkg = decision.solvable_id;
+            let watched_literal = Literal::new(decision.solvable_id, decision.value);
 
             // Propagate, iterating through the linked list of clauses that watch this
             // solvable
             let mut old_predecessor_clause_id: Option<ClauseId>;
             let mut predecessor_clause_id: Option<ClauseId> = None;
-            let mut clause_id = self.watches.first_clause_watching_solvable(pkg);
+            let mut clause_id = self
+                .watches
+                .first_clause_watching_literal(watched_literal)
+                .unwrap_or(ClauseId::null());
             while !clause_id.is_null() {
                 debug_assert!(
                     predecessor_clause_id != Some(clause_id),
                     "Linked list is circular!"
                 );
 
                 // Get mutable access to both clauses.
-                let mut clauses = self.clauses.borrow_mut();
-                let (predecessor_clause, clause) =
+                let (predecessor_clause_state, clause_state) =
                     if let Some(prev_clause_id) = predecessor_clause_id {
-                        let (predecessor_clause, clause) =
-                            clauses.get_two_mut(prev_clause_id, clause_id);
-                        (Some(predecessor_clause), clause)
+                        let prev_idx = prev_clause_id.to_usize();
```

Comment:
I think internally these are i32? Could we skip the usize casting and compare the inner values directly? Not important though.